src/Stratis.Bitcoin.Features.Consensus/ProvenBlockHeaders/RewindDataIndexStore.cs

```diff
@@ -0,0 +1,108 @@
+﻿using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Logging;
+using NBitcoin;
+using Stratis.Bitcoin.Features.Consensus.CoinViews;
+using Stratis.Bitcoin.Utilities;
+
+namespace Stratis.Bitcoin.Features.Consensus.ProvenBlockHeaders
+{
+    /// <inheritdoc />
+    public class RewindDataIndexStore : IRewindDataIndexStore
+    {
+        /// <summary>
+        /// The date time provider. 
+        /// </summary>
+        private readonly IDateTimeProvider dateTimeProvider;
+
+        /// <summary>
+        /// The rewind data index repository that is used to store and retrieve data from the disk.
+        /// </summary>
+        private readonly IRewindDataIndexRepository rewindDataIndexRepository;
+
+        /// <summary>
+        /// Internal cache for rewind data index. Key is a TxId + N (N is an index of output in a transaction)
+        /// and value is a rewind data index
+        /// </summary>
+        private readonly ConcurrentDictionary<string, int> items;
+
+        /// <summary>Time of the last cache flush.</summary>
+        private DateTime lastCacheFlushTime;
+
+        /// <summary>Length of the coinview cache flushing interval in seconds.</summary>
+        /// <seealso cref="lastCacheFlushTime"/>
+        public const int CacheFlushTimeIntervalSeconds = 60;
```

Comment:
Do we nee da flush interval? dont we just flush when coinview flushes?

Reply:
Yeah, we probably don't need it if it is controlled by coin view. I will remove it for now.